### ISSUE TITLE
bug(common): date pipe rounds up date when fractional second is more than 3 digits

### ISSUE THREAD

**Issue by antoinebrault** [angular/angular#37989]
Hi, 

I believe I've come across a bug in the date parsing. 
I haven't seen any indications that this is expected behavior, so I'm opening an issue here.

# 🐞 bug report

### Affected Package
@angular/common
Seems to be in every version I tested. Angular 6, 9, 10

https://github.com/angular/angular/blob/d1ea1f4c7f3358b730b0d94e65b00bc28cae279c/packages/common/src/i18n/format_date.ts#L735

### Is this a regression?
Not that I've seen.

### Description
The Date pipe has some unexpected behavior when second fractions have more than 3 digits.

ISO 8601 format is not capped on fractional second number of digits.
https://www.w3.org/TR/NOTE-datetime
https://en.wikipedia.org/wiki/ISO_8601#:~:text=There%20is%20no%20limit%20on%20the%20number%20of%20decimal%20places%20for%20the%20decimal%20fraction
https://momentjs.com/docs/#:~:text=Fractional%20Second
https://date-fns.org/v2.14.0/docs/format#:~:text=Fraction%20of%20second

## 🔬 Minimal Reproduction
https://stackblitz.com/edit/angular-date-fractional-second-test
```
  3 digits: {{ '2020-08-01T23:59:59.999' | date: 'yyyy-MM-dd' }} 
  <br>
  4 digits: {{ '2020-08-01T23:59:59.9999' | date: 'yyyy-MM-dd' }}
```

**Comment by petebacondarwin** [comment]
I am not sure of the history behind this but I think it might be by design. I note that the ISO 8601 specification says:

> There is no limit on the number of decimal places for the decimal fraction. However, the number of decimal places needs to be agreed to by the communicating parties. For example, in Microsoft SQL Server, the precision of a decimal fraction is 3, i.e., "yyyy-mm-ddThh:mm:ss[.mmm]".

In other words the number of decimal places is implementation specific. And it appears that Angular has made the decision that it only accepts 3 decimal places.

In the [Angular docs](https://angular.io/api/common/DatePipe) there is a reference to the [W3C Date and Time](https://www.w3.org/TR/NOTE-datetime) document which restricts the ISO 8601 standard to a smaller number of formats. In particular it says:

> This profile does not specify how many digits may be used to represent the decimal fraction of a second. An adopting standard that permits fractions of a second must specify both the minimum number of digits (a number greater than or equal to one) and the maximum number of digits (the maximum may be stated to be "unlimited").

That suggests that we should specify in our documentation what the maximum number of digits supported is. It is arguable that the table of "custom" formats in the `DatePipe` page implies that we only support 1, 2 and 3 digits, but I think we could/should be more clear about how we parse strings passed to `DatePipe` in order to then be rendered.

So I would say that this is not actually a bug in Angular, but instead either a documentation bug or a feature request to support more fraction digits.

**Comment by antoinebrault** [comment]
IMO there should at least be a way to customize it. I don't think Angular should force a choice of implementation by rounding up in that case. 

Even if Angular uses 3 digits, this should return the same result, not round up:
{{ '2020-08-01T23:59:59.999' | date: 'yyyy-MM-dd' }} 
{{ '2020-08-01T23:59:59.9999' | date: 'yyyy-MM-dd' }}

9999 should be parsed as 999, not 1000.

**Comment by petebacondarwin** [comment]
We looked into this a little bit more. So it turns out that there is a real constraint on this. The JavaScript `Date` object only works at the resolution of milliseconds. I.e. you cannot store sub-millisecond information in a `Date`. There is a W3C spec for high-resolution time objects: https://www.w3.org/TR/hr-time-2/#introduction. So when we parse a string into a Date object inside the `DatePipe` we must do something about sub-millisecond values.

The implementation we have is that we round to the nearest millisecond, which in most cases is the most reasonable. But I can see how on the border of midnight this could cause confusion.

I think we should make this explicit in the documentation, with the example you give as a potential confusion point.

The only way around this would be to implement a high precision date-time object internally in the `DatePipe` which would add loads of complexity and size to the core framework for a rare corner-case.

**Comment by antoinebrault** [comment]
Any reason you can't do something like this ? 
```
if (ms === 1000) {
    ms = 999;
}
```

It doesn't make sense to me that 2020-08-01T23:59:59.9999 is parsed as 
2020-08-02T00:00:00.000 instead of 
2020-08-01T23:59:59.999
because ```setUTCHours(hoursValue[, minutesValue[, secondsValue[, msValue]]]))```` is called with 1000

Mozilla doc:
msValue
Optional. A number between 0 and 999, representing the milliseconds.

**Comment by petebacondarwin** [comment]
> FYI new Date('2020-08-01T23:59:59.9999').getMilliseconds() returns 999 in IE 11, Chrome, Firefox, Edge 83

That is interesting to know... perhaps we should follow that convention.

**Comment by petebacondarwin** [comment]
It seems that the fourth digit is just ignored

```
new Date('2020-08-01T23:59:59.9229').getMilliseconds()
922
```

So perhaps we should just change the `Math.round()` to `Math.floor()` instead?

**Comment by petebacondarwin** [comment]
OK, so I think you have convinced me that this is a bug @antoinebrault  :-)

**Comment by petebacondarwin** [comment]
Some more research... as indicated by the example given earlier the ECMAScript spec defines that non-integer values passed for any of the parameters to `new Date(year, month, date, hours, minutes, seconds, ms)` calls `floor()` to remove the fractional part: https://www.ecma-international.org/ecma-262/5.1/#sec-15.9.1.11

**Comment by petebacondarwin** [comment]
So I think it would be reasonable to "fix" this, but I think it would count as a breaking change and not be released until 11.0.0
